packages/plugin-ext/src/hosted/browser/hosted-plugin.ts

```diff
@@ -436,39 +431,35 @@ export class HostedPluginSupport {
             this.managers.set(host, manager);
             toDisconnect.push(Disposable.create(() => this.managers.delete(host)));
 
-            const [extApi, globalState, workspaceState, webviewResourceRoot, webviewCspSource, defaultShell, jsonValidation] = await Promise.all([
-                this.server.getExtPluginAPI(),
-                this.pluginServer.getAllStorageValues(undefined),
-                this.pluginServer.getAllStorageValues({
-                    workspace: this.workspaceService.workspace,
-                    roots: this.workspaceService.tryGetRoots()
-                }),
-                this.webviewEnvironment.resourceRoot(),
-                this.webviewEnvironment.cspSource(),
-                this.terminalService.getDefaultShell(),
-                this.jsonSchemaStore.schemas
-            ]);
+            const extApi = await this.server.getExtPluginAPI();
             if (toDisconnect.disposed) {
                 return undefined;
             }
 
+            const mainApiProviders = new Map<string, MainPluginApiProvider>();
+            this.mainPluginApiProviders.getContributions().forEach(provider => {
+                mainApiProviders.set(provider.id, provider);
+            });
+
+            const apiInitParams: {
+                pluginApi: ExtPluginApi,
+                initParameters: any
+            }[] = [];
+
+            for (const api of extApi) {
+                const result = {
+                    pluginApi: api,
+                    initParameters: undefined
+                };
+                const mainProvider = mainApiProviders.get(api.id);
+                if (mainProvider && mainProvider.computeInitParameters) {
+                    result.initParameters = await mainProvider.computeInitParameters(rpc, this.container);
```

Comment:
we should check whether we disconnected after any async call and exit with undefined as it was before

Reply:
you mean like 
```
if (toDisconnect.disposed) {
  return undefined;
}
``